File: Services/LocationService.cs
using MebToplantiTakip.DbContexts;
using MebToplantiTakip.Dtos;
using MebToplantiTakip.Entities;
using Microsoft.EntityFrameworkCore;

namespace MebToplantiTakip.Services
{
    public class LocationService(MebToplantiTakipContext context)
    {
        public async Task<Location> AddLocation (LocationDto location)
        {
            var Location = new Location
            {
                Latitude = location.Latitude,
                Longitude = location.Longitude,
                LocationName = location.LocationName
            };
            await context.Locations.AddAsync (Location);    
            await context.SaveChangesAsync ();
            return Location;
        }

        public async Task<List<Location>> GetAllLocations()
        {
            return await context.Locations.AsNoTracking().ToListAsync();
        }

        public async Task<Location> GetLocationById(int locationId)
        {
            return await context.Locations.FindAsync(locationId);
        }

        public async Task<Location> UpdateLocation(Location location)
        {
            if (location == null || location.LocationId <= 0)
                throw new ArgumentException("Geçersiz lokasyon bilgisi");

            var existingLocation = await context.Locations.FindAsync(location.LocationId);
            if (existingLocation == null)
                throw new Exception("Lokasyon bulunamadı");

            existingLocation.LocationName = location.LocationName;
            existingLocation.Latitude = location.Latitude;
            existingLocation.Longitude = location.Longitude;

            context.Locations.Update(existingLocation);
            await context.SaveChangesAsync();
            return existingLocation;
        }

        public async Task<bool> DeleteLocation(int locationId)
        {
            var location = await context.Locations.FindAsync(locationId);
            if (location == null)
                return false;

            // Önce bu lokasyonu kullanan toplantılar var mı kontrol et
            var meetingsUsingLocation = await context.Meetings
                .Where(m => m.Location != null && m.Location.LocationId == locationId)
                .CountAsync();

            if (meetingsUsingLocation > 0)
                throw new Exception("Bu lokasyon toplantılarda kullanıldığı için silinemiyor");

            context.Locations.Remove(location);
            await context.SaveChangesAsync();
            return true;
        }

        public async Task<List<Location>> SearchLocationsByName(string locationName)
        {
            if (string.IsNullOrWhiteSpace(locationName))
                return new List<Location>();

            return await context.Locations
                .AsNoTracking()
                .Where(l => l.LocationName.Contains(locationName))
                .ToListAsync();
        }

        public async Task<List<Location>> GetLocationsByDistance(double latitude, double longitude, double maxDistanceKm)
        {
            // Basit mesafe hesaplama (Haversine formülü kullanılabilir daha doğru hesaplama için)
            return await context.Locations
                .AsNoTracking()
                .Where(l => Math.Abs(l.Latitude - latitude) <= maxDistanceKm / 111.0 && 
                           Math.Abs(l.Longitude - longitude) <= maxDistanceKm / 111.0)
                .ToListAsync();
        }

        public async Task<bool> IsLocationNameExists(string locationName, int? excludeLocationId = null)
        {
            var query = context.Locations.AsNoTracking()
                .Where(l => l.LocationName.ToLower() == locationName.ToLower());

            if (excludeLocationId.HasValue)
                query = query.Where(l => l.LocationId != excludeLocationId.Value);

            return await query.AnyAsync();
        }
    }
}

File: phpapi/src/Services/LocationService.php
<?php

namespace App\Services;

use App\Models\Location;
use Exception;

class LocationService
{
    public function getAllLocations()
    {
        return Location::all();
    }

    public function getLocationById(int $id)
    {
        return Location::find($id);
    }

    public function createLocation(array $data)
    {
        if (empty($data['LocationName'])) {
            throw new Exception('Lokasyon adı zorunludur');
        }

        if (!isset($data['Latitude']) || !isset($data['Longitude'])) {
            throw new Exception('Enlem ve boylam bilgileri zorunludur');
        }

        // Koordinat geçerliliği kontrolü
        $latitude = (float)$data['Latitude'];
        $longitude = (float)$data['Longitude'];

        if ($latitude < -90 || $latitude > 90) {
            throw new Exception('Enlem değeri -90 ile 90 arasında olmalıdır');
        }

        if ($longitude < -180 || $longitude > 180) {
            throw new Exception('Boylam değeri -180 ile 180 arasında olmalıdır');
        }

        // Lokasyon adı benzersizliği kontrolü
        if ($this->isLocationNameExists($data['LocationName'])) {
            throw new Exception('Bu lokasyon adı zaten kullanılmakta');
        }

        $locationData = [
            'LocationName' => $data['LocationName'],
            'Latitude' => $latitude,
            'Longitude' => $longitude
        ];

        return Location::create($locationData);
    }

    public function updateLocation(int $id, array $data)
    {
        $location = Location::find($id);
        
        if (!$location) {
            return null;
        }

        // Koordinat güncelleniyorsa geçerlilik kontrolü
        if (isset($data['Latitude'])) {
            $latitude = (float)$data['Latitude'];
            if ($latitude < -90 || $latitude > 90) {
                throw new Exception('Enlem değeri -90 ile 90 arasında olmalıdır');
            }
        }

        if (isset($data['Longitude'])) {
            $longitude = (float)$data['Longitude'];
            if ($longitude < -180 || $longitude > 180) {
                throw new Exception('Boylam değeri -180 ile 180 arasında olmalıdır');
            }
        }

        // Lokasyon adı güncelleniyorsa benzersizlik kontrolü
        if (isset($data['LocationName']) && $data['LocationName'] !== $location->LocationName) {
            if ($this->isLocationNameExists($data['LocationName'], $id)) {
                throw new Exception('Bu lokasyon adı zaten kullanılmakta');
            }
        }

        $allowedFields = ['LocationName', 'Latitude', 'Longitude'];
        $updateData = array_intersect_key($data, array_flip($allowedFields));

        $location->update($updateData);
        
        return $location->fresh();
    }

    public function deleteLocation(int $id): bool
    {
        $location = Location::find($id);
        
        if (!$location) {
            return false;
        }

        // Bu lokasyonu kullanan toplantılar var mı kontrol et
        $meetingCount = $location->meetings()->count();
        if ($meetingCount > 0) {
            throw new Exception('Bu lokasyon toplantılarda kullanıldığı için silinemez');
        }

        return $location->delete();
    }

    public function searchLocationsByName(string $locationName)
    {
        return Location::where('LocationName', 'LIKE', '%' . $locationName . '%')->get();
    }

    public function getLocationsByDistance(float $lat, float $lng, float $maxDistance)
    {
        $locations = Location::all();
        $nearbyLocations = [];

        foreach ($locations as $location) {
            $distance = $location->calculateDistance($lat, $lng);
            if ($distance <= $maxDistance) {
                $locationArray = $location->toArray();
                $locationArray['distance'] = round($distance, 2);
                $nearbyLocations[] = $locationArray;
            }
        }

        // Mesafeye göre sırala
        usort($nearbyLocations, function($a, $b) {
            return $a['distance'] <=> $b['distance'];
        });

        return $nearbyLocations;
    }

    public function isLocationNameExists(string $name, int $excludeId = null): bool
    {
        $query = Location::where('LocationName', $name);
        
        if ($excludeId) {
            $query->where('LocationId', '!=', $excludeId);
        }
        
        return $query->exists();
    }
} 